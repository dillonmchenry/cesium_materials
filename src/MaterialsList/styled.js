import styled from 'styled-components';

export const ListWrapper = styled.div`
    position: relative;
    width: 35%;
    max-height: 325px;
    background: #17171A;
    text-align: center;
    border: 1px solid gray;
    overflow: auto;
`;

export const MaterialWrapper = styled.div`
    display: flex;
    align-items: center;
    padding: 10px 15px;
    cursor: pointer;
    text-align: left;
    border-bottom: 1px solid gray;
    ${({ selected }) => selected && `background: #000344`};
    
    p {
        margin: 4px 0;
    }
    
    p.note {
        font-size: 12px;
    }
    
    div.circle {
        width: 30px;
        height: 30px;
        margin-right: 10px;
        border-radius: 50px;
        background: ${({ color }) => color};
    }
    
`;