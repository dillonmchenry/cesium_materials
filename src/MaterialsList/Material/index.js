import React from 'react';
import PropTypes from 'prop-types';
import {MaterialWrapper} from "../styled";

const Material = ({
    name,
    volume,
    color,
    selected,
    handleSelect,
}) => {

    return (
        <MaterialWrapper onClick={handleSelect} selected={selected} color={color}>
            <div className="circle" color={color}> </div>
            <div>
                <p>{name}</p>
                <p className="note">{volume}m3</p>
            </div>
        </MaterialWrapper>
    );
}

Material.propTypes = {
    name: PropTypes.string,
    volume: PropTypes.number,
    color: PropTypes.string,
    selected: PropTypes.bool,
    handleSelect: PropTypes.func,
}

export default Material;