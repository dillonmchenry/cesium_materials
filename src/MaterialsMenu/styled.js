import styled from 'styled-components';

export const MenuWrapper = styled.div`
    width: 100%;
    max-width: 700px;
    margin: 50px auto;
    color: white;
    background: #29292F;
    padding: 0 30px;
    
    @media (min-width: 750px) {
        max-height: 300px;
    }
    
    h2 {
        font-weight: 500;
    }
`;

export const ButtonWrapper = styled.div`
    display: flex;
    margin: 15px 0;
    
    button {
        color: white;
        cursor: pointer;
        border: 0;
        border-radius: 16px;
        padding: 7px 12px;
        margin-right: 10px;
    }
    
    .add {
        background: #0075DB;
    }
    
    .delete {
        background: #FF444C
    }
    
`;

export const MaterialGrid = styled.div`
    position: relative;
    display: flex;
    flex-direction: row;
    
    @media (max-width: 750px) {
        flex-direction: column;
    }

`;

