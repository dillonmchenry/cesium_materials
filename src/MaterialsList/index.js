import React from 'react';
import PropTypes from 'prop-types';
import { ListWrapper } from "./styled";
import Material from "./Material";

const MaterialsList = ({ materials, handleSelect, selected }) => {
    return (
        <ListWrapper>
            {materials.length === 0 ?
                <p>No materials</p> :
                materials.map((material, idx) => {
                    return (
                        <Material
                            key={idx}
                            selected={selected === idx}
                            handleSelect={() => handleSelect(idx)}
                            name={material.name}
                            volume={material.volume}
                            color={material.color}
                        />
                    )
                })
            }
        </ListWrapper>
    );
}

MaterialsList.propTypes = {
    materials: PropTypes.array.isRequired,
    handleSelect: PropTypes.func.isRequired,
    selected: PropTypes.number,
}

export default MaterialsList;