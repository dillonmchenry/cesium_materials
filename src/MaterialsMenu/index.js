import React from 'react';
import {ButtonWrapper, MaterialGrid, MenuWrapper} from "./styled";

import MaterialsList from "../MaterialsList";
import MaterialDetails from "../MaterialDetails";

class MaterialsMenu extends React.Component {
    constructor(props) {
        super(props);

        this.state = ({
            materials: [],
            selected: null,
        })
    }

    handleAddMaterial = () => {
        const materials = [...this.state.materials]

        this.setState({
            materials: [{
                name: "New Material",
                volume: 0,
                deliveryDate: null,
                color: "#ffffff",
                cost: 0.00,
            }, ...materials]
        });
    }

    handleDeleteMaterial = () => {
        const selected = this.state.selected;

        if (selected !== null) {
            const materials = [...this.state.materials]
            materials.splice(selected, 1);
            this.setState({
                materials: materials,
                selected: null,
            })
        }
    }

    handleMaterialSelection = (index) => {
        console.log("selection handled");
        this.setState({
            selected: index,
        });
    }

    handleMaterialChange = (e, field) => {
        const value = e.target.value;
        const selected = this.state.selected;
        const materials = this.state.materials;

        switch (field) {
            case "name":
                materials[selected].name = value;
            break;
            case "volume":
                materials[selected].volume = value;
            break;
            case "deliveryDate":
                materials[selected].deliveryDate = value;
            break;
            case "color":
                materials[selected].color = value;
            break;
            case "cost":
                materials[selected].cost = value;
            break;
            default:
                break;
        }
        this.setState({ materials: materials })
    }

    getSelectedMaterial = () => {
        const selected = this.state.selected;

        if (selected !== null) {
            return this.state.materials[selected];
        }
        return null;
    }

    getTotalCost = () => {
        let cost = 0.00;
        const materials = this.state.materials;

        for (let i=0; i<materials.length; i++) {
            cost += materials[i].cost*materials[i].volume;
        }
        return cost.toFixed(2);
    }

    render() {
        return (
            <MenuWrapper>
                <h2>Materials</h2>
                <ButtonWrapper>
                    <button
                        className="add"
                        onClick={this.handleAddMaterial}
                    >+ &nbsp;Add</button>
                    <button
                        className="delete"
                        onClick={this.handleDeleteMaterial}
                    >- &nbsp;Delete</button>
                </ButtonWrapper>
                <MaterialGrid>
                    <MaterialsList
                        materials={this.state.materials}
                        selected={this.state.selected}
                        handleSelect={this.handleMaterialSelection}
                    />
                    <MaterialDetails
                        material={this.getSelectedMaterial()}
                        handleChange={this.handleMaterialChange}
                    />
                </MaterialGrid>
                <p className="total-cost">Total Cost: &nbsp;&nbsp; ${this.getTotalCost()}</p>
            </MenuWrapper>
        )
    }
}

export default MaterialsMenu;