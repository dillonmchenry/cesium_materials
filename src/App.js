import React from "react";
import MaterialsMenu from "./MaterialsMenu";

function App() {
  return (
    <div className="App">
      <MaterialsMenu />
    </div>
  );
}

export default App;
