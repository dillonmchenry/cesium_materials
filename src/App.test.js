import React from "react";
import { render, screen } from '@testing-library/react';
import App from './App';
import MaterialsMenu from "./MaterialsMenu";

import Enzyme, { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import MaterialsList from "./MaterialsList";

Enzyme.configure({ adapter: new Adapter() })


test('renders menu correctly', () => {
  const rendered = shallow(<MaterialsMenu />);
  expect(toJson(rendered)).toMatchSnapshot();
});

test('adding a material works correctly', () => {
  const rendered = shallow(<MaterialsMenu />);
  rendered.find('.add').simulate('click');
  const list = rendered.find(MaterialsList);
  expect(list.getElement().props.materials.length).toEqual(1);
})

test('total cost renders correctly', () => {
  const rendered = shallow(<MaterialsMenu />);
  rendered.setState({
    materials: [
      {
        volume: 30,
        cost: 13.65,
      },
      {
        volume: 10,
        cost: 543.12,
      }
    ]
  })
  const totalcost = rendered.find('.total-cost');
  expect(totalcost.getElement().props.children[1]).toEqual("5840.70");
});


