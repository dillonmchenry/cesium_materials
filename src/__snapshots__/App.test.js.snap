// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders menu correctly 1`] = `
<styled.div>
  <h2>
    Materials
  </h2>
  <styled.div>
    <button
      className="add"
      onClick={[Function]}
    >
      +  Add
    </button>
    <button
      className="delete"
      onClick={[Function]}
    >
      -  Delete
    </button>
  </styled.div>
  <styled.div>
    <MaterialsList
      handleSelect={[Function]}
      materials={Array []}
      selected={null}
    />
    <MaterialDetails
      handleChange={[Function]}
      material={null}
    />
  </styled.div>
  <p
    className="total-cost"
  >
    Total Cost:    $
    0.00
  </p>
</styled.div>
`;
